src/irr/asset/CCPUMeshPacker.h

```diff
@@ -155,70 +165,123 @@ ReservedAllocationMeshBuffers CCPUMeshPacker<MDIStructType>::alloc(const Iterato
 }
 
 template <typename MDIStructType>
-void CCPUMeshPacker<MDIStructType>::createOutputBuffers()
+void CCPUMeshPacker<MDIStructType>::instantiateDataStorage()
 {
-	if (outputBuffer.indexBuffer.buffer != nullptr)
-		return;
-
 	//TODO: redo after safe_shrink fix
-	outputBuffer.MDIDataBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(m_allocParams.MDIDataBuffSupportedSizeInBytes);
-	outputBuffer.indexBuffer.buffer = core::make_smart_refctd_ptr<ICPUBuffer>(m_allocParams.indexBuffSupportedSizeInBytes);
-	outputBuffer.perVertexVtxBuffer.buffer = core::make_smart_refctd_ptr<ICPUBuffer>(m_allocParams.vertexBuffSupportedSizeInBytes);
+	outputBuffer.MDIDataBuffer = core::make_smart_refctd_ptr<ICPUBuffer>(m_allocParams.MDIDataBuffSupportedCnt * sizeof(MDIStructType));
+	outputBuffer.indexBuffer.buffer = core::make_smart_refctd_ptr<ICPUBuffer>(m_allocParams.indexBuffSupportedCnt * sizeof(uint16_t));
+
+	core::smart_refctd_ptr<ICPUBuffer> unifiedVtxBuff = core::make_smart_refctd_ptr<ICPUBuffer>(m_allocParams.vertexBuffSupportedCnt * m_vtxSize);
+	core::smart_refctd_ptr<ICPUBuffer> unifiedInsBuff = core::make_smart_refctd_ptr<ICPUBuffer>(m_allocParams.perInstanceVertexBuffSupportedCnt * m_perInstVtxSize);
+
+	outputBuffer.vertexInputParams = m_outVtxInputParams;
+
+	//divide unified vtx buffers
+	//proportions: sizeOfAttr1 : sizeOfAttr2 : ... : sizeOfAttrN
+	std::array<uint32_t, SVertexInputParams::MAX_ATTR_BUF_BINDING_COUNT> attrSizeArray;
+
+	uint32_t perVtxAttrSizeSum = 0u;
+	uint32_t perInsAttrSizeSum = 0u;
+	uint16_t activeAttribCnt = 0u;
+
+	for (uint16_t attrBit = 0x0001, location = 0; location < SVertexInputParams::MAX_ATTR_BUF_BINDING_COUNT; attrBit <<= 1, location++)
+	{
+		auto attrib = m_outVtxInputParams.attributes[location];
+		auto binding = m_outVtxInputParams.bindings[attrib.binding];
+
+		if (!(attrBit & m_outVtxInputParams.enabledAttribFlags))
+		{
+			attrSizeArray[location] = 0u;
+			continue;
+		}
+		
+		attrSizeArray[location] = asset::getTexelOrBlockBytesize(static_cast<E_FORMAT>(attrib.format));
+
+		if (binding.inputRate == EVIR_PER_VERTEX)
+			perVtxAttrSizeSum += attrSizeArray[location];
+		else
+			perInsAttrSizeSum += attrSizeArray[location];
+
+		activeAttribCnt++;
+	}
+	
+	uint32_t perVtxUnitVal = (m_allocParams.vertexBuffSupportedCnt * m_vtxSize + perVtxAttrSizeSum - 1) / perVtxAttrSizeSum; //round up??
+	uint32_t perInsUnitVal = 0u;
 	if(m_perInstVtxSize)
-		outputBuffer.perInstanceVtxBuffer.buffer = core::make_smart_refctd_ptr<ICPUBuffer>(m_allocParams.perInstanceVertexBuffSupportedSizeInBytes);
-	outputBuffer.perVertexVtxBuffer.offset = 0u;
-	outputBuffer.perInstanceVtxBuffer.offset = 1u;
+		perInsUnitVal = (m_allocParams.perInstanceVertexBuffSupportedCnt * m_vtxSize + perInsAttrSizeSum - 1) / perInsAttrSizeSum; //round up??
```

Comment:
you'll fail if you don't handle vertex and index buffers separately, you need two separate buffers that you partition separately

Reply:
why do you think rounding is necessary?